packages/shared/pkg/storage/storage_google.go

```diff
@@ -210,23 +217,61 @@ func (g *GCPBucketStorageObjectProvider) WriteTo(dst io.Writer) (int64, error) {
 }
 
 func (g *GCPBucketStorageObjectProvider) WriteFromFileSystem(path string) error {
-	cmd := exec.CommandContext(
-		g.ctx,
-		"gcloud",
-		"storage",
-		"cp",
-		"--verbosity",
-		"error",
-		path,
-		fmt.Sprintf("gs://%s/%s", g.storage.bucket.BucketName(), g.path),
-	)
+	upload := func() error {
+		systemRunArgs := []string{
+			"--user",
+			"--scope",
+		}
+		if g.limiter != nil {
+			uploadLimiter := g.limiter.GCloudUploadLimiter()
+			if uploadLimiter != nil {
+				semaphoreErr := uploadLimiter.Acquire(g.ctx, 1)
+				if semaphoreErr != nil {
+					return fmt.Errorf("failed to acquire semaphore: %w", semaphoreErr)
+				}
+				defer uploadLimiter.Release(1)
+			}
+
+			systemRunArgs = append(systemRunArgs, g.limiter.GCloudCmdLimits(path)...)
+		}
 
-	output, err := cmd.CombinedOutput()
-	if err != nil {
-		return fmt.Errorf("failed to upload file to GCS: %w\n%s", err, string(output))
+		args := append(systemRunArgs,
+			"gcloud",
+			"storage",
+			"cp",
+			"--verbosity", "error",
+			path,
+			fmt.Sprintf("gs://%s/%s", g.storage.bucket.BucketName(), g.path),
+		)
+
+		cmd := exec.CommandContext(
+			g.ctx,
+			"systemd-run",
+			args...,
+		)
+
+		output, err := cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("failed to upload file to GCS: %w\n%s", err, string(output))
+		}
+
+		return nil
 	}
 
-	return nil
+	for range gcloudMaxRetries {
+		err := upload()
+		if err != nil {
+			// Failed to upload file, retrying.
+			zap.L().Warn("Failed to upload file to GCS, retrying", zap.Error(err), zap.String("path", g.path))
+
+			continue
+		}
+
+		// Files was successfully uploaded
+		return nil
+	}
+
+	return fmt.Errorf("failed to upload file to GCS after %d retries", gcloudMaxRetries)
```

Comment:
Should we include the last error as well? It would probably make debugging issues easier; you wouldn't have to go hunting for the last error.